crates/sui-move-build/src/lib.rs

```diff
@@ -286,6 +297,16 @@ pub fn build_from_resolution_graph(
         }
         Ok((package, fn_info)) => (package, fn_info),
     };
+
+    let deps: BTreeSet<_> = package
+        .all_compiled_units()
+        .filter_map(|x| {
+            let address = x.address.into_inner();
+            (address != AccountAddress::ZERO).then_some(address)
+        })
+        .collect();
+    dependency_ids.published.retain(|_, id| deps.contains(id));
```

Comment:
This is the main workhorse. It takes all the compiled units, which are all compiled units for this package in transitive dependencies, and then keeps in the dependency list only those that are present in the modules' dependencies.